File: rust/2015/day14a/src/main.rs
use std::collections::HashMap;
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::process;

use regex::Regex;

const TIME: u32 = 2503;

struct Reindeer {
    velocity: u32,
    flying_time: u32,
    rest_time: u32,
}

type ReindeerTable = HashMap<String, Reindeer>;

enum ReindeerAction {
    Flying(u32),
    Resting(u32),
}

struct ReindeerState {
    action: ReindeerAction,
    position: u32,
}

type StateTable = HashMap<String, ReindeerState>;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn parse_file(filename: &str) -> io::Result<ReindeerTable> {
    let mut table: ReindeerTable = HashMap::new();
    let re = Regex::new(
        r"(\w+) can fly (\d+) km/s for (\d+) seconds, but then must rest for (\d+) seconds.",
    )
    .unwrap();
    let file = File::open(filename)?;
    let reader = io::BufReader::new(file);
    for line_result in reader.lines() {
        let line = line_result?;
        if let Some(caps) = re.captures(&line) {
            let name = String::from(caps.get(1).unwrap().as_str());
            let reindeer = Reindeer {
                velocity: caps.get(2).unwrap().as_str().parse::<u32>().unwrap_or(0),
                flying_time: caps.get(3).unwrap().as_str().parse::<u32>().unwrap_or(0),
                rest_time: caps.get(4).unwrap().as_str().parse::<u32>().unwrap_or(0),
            };
            table.insert(name, reindeer);
        }
    }
    Ok(table)
}

fn init_race(table: &ReindeerTable) -> StateTable {
    let mut state: StateTable = HashMap::new();
    table.iter().for_each(|(name, data)| {
        state.insert(
            name.clone(),
            ReindeerState {
                action: ReindeerAction::Flying(data.flying_time - 1),
                position: 0,
            },
        );
    });
    state
}

fn process_file(filename: &str) -> io::Result<u32> {
    let table = parse_file(filename)?;
    let mut state = init_race(&table);
    for _ in 0..TIME {
        for (name, curr_state) in state.iter_mut() {
            let reindeer = table.get(name).unwrap();
            match curr_state.action {
                ReindeerAction::Flying(n) => {
                    curr_state.position += reindeer.velocity;
                    if n > 0 {
                        curr_state.action = ReindeerAction::Flying(n - 1);
                    } else {
                        curr_state.action = ReindeerAction::Resting(reindeer.rest_time - 1);
                    }
                }
                ReindeerAction::Resting(n) => {
                    if n > 0 {
                        curr_state.action = ReindeerAction::Resting(n - 1);
                    } else {
                        curr_state.action = ReindeerAction::Flying(reindeer.flying_time - 1);
                    }
                }
            }
        }
    }
    Ok(state.values().max_by_key(|v| v.position).unwrap().position)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    match process_file(filename) {
        Ok(result) => println!("result = {result}"),
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}

File: haskell/2015/day14a/src/Main.hs
module Main (main) where

import Control.Monad
import Control.Monad.State
import Data.Int (Int32)
import qualified Data.Map.Strict as Map
import System.Environment (getArgs, getProgName)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)
import Text.Parsec hiding (State)
import Text.Parsec.String
import Control.DeepSeq
import System.Clock

data ReindeerAction = Flying Int
                    | Resting Int
                    deriving (Eq, Show)

type ReindeerState = (Int, Int, Int, Int, ReindeerAction)

type Race = Map.Map String ReindeerState

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <input file>"
    exitFailure

file :: Parser [(String, Int, Int, Int)]
file = line `endBy` newline

line :: Parser (String, Int, Int, Int)
line = do
       reindeer <- many1 letter
       _ <- string " can fly "
       velo <- many1 digit
       _ <- string " km/s for "
       flying <- many1 digit
       _ <- string " seconds, but then must rest for "
       resting <- many1 digit
       _ <- string " seconds."
       return (reindeer, read velo, read flying, read resting)

initRace :: [(String, Int, Int, Int)] -> Race
initRace entries =
    Map.fromList [(name, (0, velo, flying, resting, Flying flying)) | (name, velo, flying, resting) <- entries]

runStep :: State Race ()
runStep = do
          current <- get
          let next = Map.map tick current
          put next
  where
    tick :: ReindeerState -> ReindeerState
    tick (pos, velo, flying, resting, Flying timeLeft)
      | timeLeft == 0 = (pos, velo, flying, resting, Resting (resting - 1))
      | otherwise     = (pos + velo, velo, flying, resting, Flying (timeLeft - 1))
    tick (pos, velo, flying, resting, Resting timeLeft)
      | timeLeft == 0 = (pos + velo, velo, flying, resting, Flying (flying - 1))
      | otherwise     = (pos, velo, flying, resting, Resting (timeLeft - 1))

simulate :: Int -> Race -> Race
simulate n race = execState (replicateM_ n runStep) race

process :: String -> Int32
process content =
    case parse file "" content of
        Left err -> error (show err)
        Right entries ->
            let race = initRace entries
                final = simulate 2503 race
            in fromIntegral $ maximum [pos | (_, (pos, _, _, _, _)) <- Map.toList final]


showTime :: TimeSpec -> String
showTime elapsed =
    let ns = fromIntegral (toNanoSecs elapsed) :: Double
    in if ns < 1000
       then show ns ++ " ns"
       else if ns < 1000000
       then show (ns / 1000.0) ++ " μs"
       else if ns < 1000000000
            then show (ns / 1000000.0) ++ " ms"
            else show (ns / 1000000000.0) ++ " s"

main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [filename] -> do
            start <- getTime Monotonic
            content <- readFile filename
            let result = process content
            result `deepseq` return ()
            end <- getTime Monotonic
            let elapsed = diffTimeSpec start end
            putStrLn $ "result = " ++ show result
            putStrLn $ "elapsed time: " ++ showTime elapsed
        _ -> usage progname
